docs/topics/compatibility-guide-18.md

```diff
@@ -0,0 +1,623 @@
+
+
+### Prevent implicit coercions between different numeric types
+
+> **Issue**: [KT-48645](https://youtrack.jetbrains.com/issue/KT-48645)
+>
+> **Component**: Kotlin/JVM
+>
+> **Incompatible change type**: behavioral
+>
+> **Short summary**: Kotlin will avoid converting numeric values automatically to a primitive numeric type where only a downcast to that type was needed semantically
+>
+> **Deprecation cycle**:
+>
+> - < 1.5.30: the old behavior in all affected cases
+> - 1.5.30: fix the downcast behavior in generated property delegate accessors,
+>   `-Xuse-old-backend` can be used to temporarily revert to the pre-1.5.30 fix behavior
+> - \>= 1.9: fix the downcast behavior in other affected cases
+
+### Make private constructors of sealed classes really private
+
+> **Issue**: [KT-44866](https://youtrack.jetbrains.com/issue/KT-44866)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: After relaxing restrictions on where the inheritors of sealed classes can be declared in the project structure,
+> the default visibility of sealed class constructors became protected. However, until 1.8, Kotlin still allowed calling
+> explicitly declared private constructors a sealed class outside the class scope.
+>
+> **Deprecation cycle**:
+>
+> - 1.6.20: report a warning (or an error in the progressive mode) when private constructor of a sealed class 
+>   is called outside of that class
+> - 1.8.0: use default visibility rules for private constructors (call can be resolved to it only if it is inside corresponding class),
+>   the old behavior can be brought back temporarily by specifying `-XXLanguage:-UseConsistentRulesForPrivateConstructorsOfSealedClasses`
+>   compiler argument
+
+### Prohibit using operator `==` on incompatible numeric types in builder inference context
+
+> **Issue**: [KT-45508](https://youtrack.jetbrains.com/issue/KT-45508)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Kotlin 1.8 will prohibit using the operator `==` on incompatible numeric types, for example, `Int` and `Long`,
+> in the scope of builder inference lambda functions, the same way as it does currently in other contexts
+>
+> **Deprecation cycle**:
+>
+> - 1.6.20: report a warning (or an error in the progressive mode) when the operator `==` is used on incompatible numeric types
+> - 1.8.0: raise the warning to an error,
+>   `-XXLanguage:-ProperEqualityChecksInBuilderInferenceCalls` can be used to temporarily revert to the pre-1.8 behavior
+
+### Prohibit `if` without `else` and non-exhaustive `when` in right hand side of elvis operator
+
+> **Issue**: [KT-44705](https://youtrack.jetbrains.com/issue/KT-44705)
+>
+> **Component**: Core language
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Kotlin 1.8 will prohibit using `if` operator without `else` branch or non-exhaustive `when`
```

Comment:
```suggestion
> **Short summary**: Kotlin 1.8 will prohibit using the `if` operator without an `else` branch or a non-exhaustive `when`
```

Reply:
should read logically like "(if without else) or (non-exhaustive when)"

Reply:
Do articles change the meaning?

Reply:
Not sure whe "the if" and "a when". Maybe also swap them, so it will be clear that "or" is not bound to "else"?

Reply:
`if` is the operator's name. When we use a name + noun, then we put the "the" (😁) article, for example, the `Object` class. So, that's why `the if operator`.
When we talk about branches, there can be any number of them. That's why it's some branch.

When we talk about non-exhaustive `when`, we mean some code with non-exhaustive `when`; also we have exhaustive whens in our universe.

***
And we should swap them, yes, to show that a non-exhaustive `when` is not bound to "else".

Reply:
```suggestion
> **Short summary**: Kotlin 1.8 will prohibit using  a non-exhaustive `when` or the `if` expression without an `else` branch
```